rfc/0008-builtin-blocktypes/README.md

```diff
@@ -0,0 +1,212 @@
+| Responsible | `@felix-oq`          |
+<!-- 
+  Status Overview:
+  - DRAFT: The RFC is not ready for a review and currently under change. Feel free to already ask for feedback on the structure and contents at this stage.
+  - DISCUSSION: The RFC is open for discussion. Usually, we open a PR to trigger discussions.
+  - ACCEPTED: The RFC was accepted. Create issues to prepare implementation of the RFC.
+  - REJECTED: The RFC was rejected. If another revision emerges, switch to status DRAFT.
+-->
+
+## Summary
+
+Blocktypes that are built into the language are declared syntactically by denoting IO types and their properties.
+This change does not change the semantics, but it can serve as a foundation for composite block types in an upcoming RFC.
+
+## Motivation
+
+Currently, all blocktypes only exist implicitly in the language.
+To make their definition more explicit, we should support the declaration of builtin blocktypes (i.e. blocktypes that are built into the language).
+This allows users to look up IO types and properties without having to open the documentation pages.
+
+## Explanation
+
+- `builtin` keyword to indicate that the blocktype is built into the language
+  - Produces an error if the blocktype is unknown
+- `input` / `output` keywords to type the input and output of the blocktype
+  - If omitted, a blocktype has no input or output
+- `property` keyword to define a property with a name and a type
+  - Optionally, a default value can be assigned
+- Adds new keywords for existing property types:
+  - Valuetypes: `regex`, `cellrange` / `row` / `column` / `cell`, `valuetype-assignment`
+  - Typed collections: `collection<type>`
+    - Nested collections are not supported for now
+  - IO types: `File`, `FileSystem`, `TextFile`, `Sheet`, `Table`
+
+See the following section for concrete code examples.
+
+### Syntax for current block types
+
+<details>
+<summary>Click to show</summary>
+
+```jayvee
+builtin blocktype HttpExtractor {
+    property url oftype text;
+
+    output oftype File;
+}
+
+builtin blocktype ArchiveInterpreter {
+    input oftype File;
+
+    property archiveType oftype text;
+
+    output oftype FileSystem;
+}
+
+builtin blocktype FilePicker {
+    input oftype FileSystem;
+
+    property path oftype text;
+
+    output oftype File;
+}
+
+builtin blocktype TextFileInterpreter {
+    input oftype File;
+
+    property encoding oftype text;
+    property lineBreak oftype regex;
+
+    output oftype TextFile;
+}
+
+builtin blocktype TextLineDeleter {
+    input oftype TextFile;
+
+    property lines oftype collection<integer>;
+
+    output oftype TextFile;
+}
+
+builtin blocktype TextRangeSelector {
+    input oftype TextFile;
+
+    property lineFrom oftype integer;
+    property lineTo oftype integer;
+
+    output oftype TextFile;
+}
+
+builtin blocktype CSVInterpreter {
+    input oftype TextFile;
+
+    property delimiter oftype text: ",";
```

Comment:
I don't really like this, I don't think this makes it clear that it is a default parameter. Can we do: property delimiter oftype text default ","?

Reply:
I think the syntax fits well, because it uses our assignment operator `:` to assign a default value to a property (similar to how values are assigned to properties in blocks). What about generally changing `:` to `=` for assignments, so it is more obvious what happens?

Reply:
I disagree a lot with that, the left hand expression of the colon is very different between a property in a block (where it is just the property name) and here (where it is a longish description of a property that will be in the block. I think the character used is irrelevant for that concept, I just do not see a value assignment here. In a block, the left hand of the assignment solely refers to a property name, here, without the default value assignment the line is a long definition of a property with attributes (the name delimiter and a supertype).

This will also be the only value assignment in the blocktype syntax (input/outputs do not use it).

Reply:
I think Felix' approach is more imperative (property declaration + initialization / value assignment) coming from the classic notion of programming languages where you can initialize object properties already in the class, while Philip's approach is more declarative (declares property and its default value) where the object does the actual initialization incorporating the declared default values.

The imperative approach is closer to other programming languages while the declarative approach is easier for non-programmers. Personally, I'd also lean towards the declarative approach (`default` keyword) but I understand the beauty of the imperative approach as well as it leads to a less bloated language.

In general, we should discuss whether human-readability / suitability for novices is one of our core USPs or not. This should also guide us in design decisions like this one.

Reply:
I can not follow that, my point is that there is a fundamental difference between assigning a value to a property e.g. in our blocks:

```
.... {
    propertyName: value;
}
``` 

And here, where we do not have a block that has propertyNames with value assignments but a definition of a propertytype that, to be more readable, we do inline:

```
.... {
    property delimiter oftype text;
}
``` 

You could write this, following our usual syntax and not inline, as:

```
property {
    name: delimiter;
    supertype: text;
}
```

Then my point is, the default value is a property of the propertytype:

```
property {
    name: delimiter;
    supertype: text;
    defaultValue: ","
}
```

and not a value assignment, therefor the syntax of having a value assigment to the inline propertytype definition makes no sense.

Note that in all of these cases we have not imperatively gave any machine any commands but only described what properties the propertytype has.

Reply:
I think the other viewpoint is the following. Let's assume we implement this in Java, then it would look like this:

```java
class MyNewBlockType extends AbstractBlockType {
    string delimiter = ","; // declare and initialize with default value
    
    // aliternative: declaration and initialization separately
    string delimiter;
    
    MyNewBlockType() {
      delimiter = ",";
    }
}

var newBlock = new MyNewBlockType();
newBlock.delimiter = ";"; // overwrite default value
```

It is close to the object-oriented/imperative programming paradigm instead of the declarative one.
Obviously, if we look into the meta-model, there you would have a class like this or so.

```java
class Attribute<T> {
    string identifier;
    T valuetype; // dunno which type fits here
    T defaultValue;
}
```

So maybe we are just discussing what level we want to present to the Jayvee user?

Reply:
Agreed on assignment syntax in peer debriefing meeting.